XXX-staged-require-with-lifecycle-hooks.md

```diff
@@ -0,0 +1,330 @@
+
+### `ModuleRequest`
+
+##### `moduleRequest.parent`
+
+The parent module which initiated the `require()` request.
+
+##### `moduleRequest.inputPath`
+
+The input string to the `require()` call or `import` statement which initiated
+this module request.
+
+##### `moduleRequest.resolvedPath`
+
+The fully resolved file path, as determined by the `require.resolvers` handler.
+This would be added to the module request object during the `resolver` stage.
+
+##### `moduleRequest.contents`
+
+The content string (or buffer?) of the module. This field is populated during
+the `loader` stage.
+
+##### `moduleRequest.module`
+
+The final module object, built by the `compiler` stage.
+
+## Hook-based
+
+The idea here is to go for something similar to `Module._extensions`, but with
+three stages: resolving, loading and compiling.
+
+### `require.resolvers`
+
+It is the responsibility of `require.resolvers` extension handlers to receive
+the in-process `Module` object and use the metadata on it to figure out how to
+resolve a load. For `require()` relative location loading, it'd simply look at
+the `parent` and resolve the input string of the `require()` call relative to
+the `parent` module.
+
+This would roughly correspond to the current `Module._resolveLookupPaths`
+implementation, but it should instead receive the module request object which
+it will modify in-place.
+
+The first resolver to return a non-null value would decide the `resolvedPath`
+value in subsequent stages.
+
+```js
+const mockedPaths = {
+  'foo.js': path.join(__dirname, 'mocked-foo.js')
+}
+
+require.resolvers.push(({ inputPath }) => {
+  if (Object.keys(mockedPaths).includes(inputPath)) {
+    return mockedPaths[inputPath]
+  }
+})
+```
+
+### `require.loaders`
+
+The `require.loaders` extension handlers receive the `Module` object after the
+`resolver` stage and use the resolved path to attempt to load the contents of
+the module file.
+
+Returning `null` would skip to the next loader in the pipeline. If no loader
+returns a value, a `MODULE_NOT_FOUND` error is thrown.
+
+```js
+require.loaders.push(({ resolvedPath }) => {
+  if (!/\.js$/.test(resolvedPath)) return
+  try {
+    return fs.readFileSync(resolvedPath).toString()
+  } catch (err) {
+    if (err.code === 'ENOENT') return
+    throw new Error(`module load error: ${err.message}`)
+  }
+})
+```
+
+NOTE: How native modules would be handled in the three-stage design is a bit
+unknown, as they don't have a textual intermediate representation, they are
+loaded directly from the file path to a `.node` file to a JS object in C++ via
+`process.dlopen`. It might make sense for native modules to have a `compiler`
+and not a `loader`.
+
+### `require.transformers`
+
+The `require.transformers` extension handlers receive the `Module` object after
+the `loader` stage and alter the text content of the module.
+
+```js
+require.transformers.push(({ contents, resolvedPath }) => {
+  if (!/\.js$/.test(resolvedPath)) return
```

Comment:
shouldn't it `return contents`? Or is returning `undefined` the same as not modifying the contents?